File: app/src/main/java/net/daverix/ajvm/jvm/IndexedDLoadOperation.java
package net.daverix.ajvm.jvm;


import net.daverix.ajvm.Frame;
import net.daverix.ajvm.ByteCodeReader;

import java.io.IOException;

public class IndexedDLoadOperation implements ByteCodeOperation {
    private final int index;

    public IndexedDLoadOperation(int index) {
        this.index = index;
    }

    @Override
    public void execute(ByteCodeReader reader,
                        int indexOfBytecode,
                        Frame currentFrame) throws IOException {
        Object variable = currentFrame.getLocalVariable(index);
        if (!(variable instanceof Double))
            throw new IllegalStateException("variable " + variable + " is not a Double");
        
        currentFrame.push(variable);
    }
}

File: app/src/main/java/net/daverix/ajvm/operation/IndexedDLoadOperation.kt
package net.daverix.ajvm.operation


import net.daverix.ajvm.Frame
import net.daverix.ajvm.ByteCodeReader

import java.io.IOException

class IndexedDLoadOperation(private val index: Int) : ByteCodeOperation {

    @Throws(IOException::class)
    override fun execute(reader: ByteCodeReader,
                         indexOfBytecode: Int,
                         currentFrame: Frame) {
        val variable = currentFrame.getLocalVariable(index)
        if (variable !is Double)
            throw IllegalStateException("variable $variable is not a Double")

        currentFrame.push(variable)
    }
}
